### PR TITLE
gh-111062: Reusable Ubuntu build that supports free-threaded mode as the conditional CI

### PR DESCRIPTION
**Pull request by NCLI**
* Issue: gh-111062

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -595,6 +556,7 @@ jobs:
       with:
         allowed-failures: >-
           build_macos,
+          build_ubuntu_nogil,
```
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
          build_ubuntu_free_threaded,
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -617,6 +579,7 @@ jobs:
             build_win_arm64,
             build_macos,
             build_ubuntu,
+            build_ubuntu_nogil,
```
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
            build_ubuntu_free_threaded,
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -248,70 +248,30 @@ jobs:
...
-      run: make pythoninfo
-    - name: Remount sources writable for tests
-      # some tests write to srcdir, lack of pyc files slows down testing
-      run: sudo mount $CPYTHON_RO_SRCDIR -oremount,rw
-    - name: Tests
-      working-directory: ${{ env.CPYTHON_BUILDDIR }}
-      run: xvfb-run make test
+
+  build_ubuntu_nogil:
+    name: 'Ubuntu GIL disabled'
```
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
    name: 'Ubuntu (free-threaded)'
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -248,70 +248,30 @@ jobs:
...
-      working-directory: ${{ env.CPYTHON_BUILDDIR }}
-      run: make pythoninfo
-    - name: Remount sources writable for tests
-      # some tests write to srcdir, lack of pyc files slows down testing
-      run: sudo mount $CPYTHON_RO_SRCDIR -oremount,rw
-    - name: Tests
-      working-directory: ${{ env.CPYTHON_BUILDDIR }}
-      run: xvfb-run make test
+
+  build_ubuntu_nogil:
```
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
  build_ubuntu_free_threaded:
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -582,6 +542,7 @@ jobs:
     - build_win_arm64
     - build_macos
     - build_ubuntu
+    - build_ubuntu_nogil
```
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
    - build_ubuntu_free_threaded
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -248,70 +248,30 @@ jobs:
...
-      # some tests write to srcdir, lack of pyc files slows down testing
-      run: sudo mount $CPYTHON_RO_SRCDIR -oremount,rw
-    - name: Tests
-      working-directory: ${{ env.CPYTHON_BUILDDIR }}
-      run: xvfb-run make test
+
+  build_ubuntu_free_threaded:
+    name: 'Ubuntu (free-threaded)'
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
```
Comment (corona10) [inline .github/workflows/build.yml]: We should check that the label is attached   or not.

**Review by corona10 — CHANGES_REQUESTED** [review]
Please skip the test if the label is not attached
https://github.com/python/cpython/pull/111452/files#r1375408558

File: .github/workflows/build.yml (hunk 6 above)
Comment (corona10) [inline .github/workflows/build.yml]:
```suggestion
    if: needs.check_source.outputs.run_tests == 'true' && contains(github.event.pull_request.labels.*.name, 'topic-free-threaded')
```

**Review by corona10 — APPROVED** [review]
Now it looks good to me
@hugovk @ezio-melotti  Can you please take a look?

**Comment by hugovk** [comment]
I'll add the https://github.com/python/cpython/labels/topic-free-threaded label to check if it triggers.

**Comment by hugovk** [comment]
The label would need to have been added before the _workflow_ runs, which means you need to add it when drafting the PR or do it very quickly.

Or we need to add triggers similar to this:

https://github.com/python/cpython/blob/fa35b9e89b2e207fc8bae9eb0284260d0d922e7a/.github/workflows/require-pr-label.yml#L3-L5

But that adds extra complexity because we don't want to re-trigger all the other CI jobs.

---

More to the point: do we really need to limit this via a label?

Should we always run it?

After all, we want to make sure there are no regressions for both the GIL and free-threaded builds, whether or not someone considers their PR to somehow related free-threaded.

**Comment by corona10** [comment]
> The label would need to have been added before the workflow runs, which means you need to add it when drafting the PR > or do it very quickly.
> 
> Or we need to add triggers similar to this:
> More to the point: do we really need to limit this via a label?
>  Should we always run it?

I know this issue, but `topiuc-free-thread` works are not a big portion of over all PRs so I think that it can be taken into account anyway at this moment. In the near future it can be run anyway once the `free-thread` version becomes stable but looks like not this moment. :)

**Comment by corona10** [comment]
> The label would need to have been added before the workflow runs, which means you need to add it when drafting the PR or do it very quickly.

And it's intended design at this moment, or we need to improve it at later

**Comment by corona10** [comment]
And simply re-open the PR will works if the person miss the timing of the labeling

**Review by hugovk — APPROVED** [review]
Thanks, let's merge this.

We can remove the label-handling later when we're all ready for it.

Let's also consider the YML file names later, once the Windows one is also done.

And at the same time we can see about naming the "build_x" bit (the third part in "Tests / OS / build_x") in a nicer and more consistent way:

image

PS We'll need to backport all of these PRs to 3.11/3.12 -- but obviously not the free threaded part. And we can take care of that later too.

**Comment by NCLI** [comment]
Thanks for the help @corona10 :-)

**Pull request merged**